personalrobotics/ada_feeding
# Implement `servo_until` and `servo_until_pose`

# Description

Linked with [`ada_ros2`#33](https://github.com/personalrobotics/ada_ros2/pull/33) and [PRL's fork of moveit2 (branch: `amaln/main`)](https://github.com/personalrobotics/moveit2/tree/amaln/main).

Currently, we can only servo for a duration. This PR extends that by implementing an idiom, `servo_until`` allowing us to servo until an arbitrary condition is met. It further implements a version of that idiom, `servo_until_pose`, that servos until the end effector is at a target pose.

# Testing procedure

Pull the linked repositories (see above) and rebuild. Set up the code as in the README. Pull [`ada_ros2`#30](https://github.com/personalrobotics/ada_ros2/pull/30).

    - [x] `ros2 action send_goal /MoveToStagingConfiguration ada_feeding_msgs/action/MoveTo "{}" --feedback`
    - [x] `ros2 action send_goal /MoveToMouth ada_feeding_msgs/action/MoveToMouth "{}" --feedback`. Verify that it smoothly servos to 2.5cm in front of the mouth.
    - [x] `ros2 action send_goal /MoveToStagingConfiguration ada_feeding_msgs/action/MoveTo "{}" --feedback`
    - [x] `ros2 action send_goal /MoveToMouth ada_feeding_msgs/action/MoveToMouth "{}" --feedback`. Verify that it smoothly servos to 2.5cm in front of the mouth.
    - [x] `ros2 action send_goal /MoveFromMouthToStagingConfiguration ada_feeding_msgs/action/MoveTo "{}" --feedback`

# Before opening a pull request

# Before Merging
